Session and path message refactor

tewinget opened the pull request:
This adds authentication to path and session messages, and encryption to session control messages and path control responses.  All traffic should now be correctly encrypted and authenticated.

tewinget committed:
temp commit, WIP session message refactor

- path messages refactored to include MAC, replies as well
- session messages refactored to include MAC, replies as well
- session tags refactored to have one in each direction per session,
  chosen by the recipient for that direction
- session construction/init refactored w.r.t. session control message and tag changes

TODO:
- a few session messages to re-implement
- need to rebase on some upstream changes
- testing, of course

tewinget committed:
fix path message size calc / assertion fail

tewinget committed:
correct log message

tewinget committed:
fix path control response encryption

tewinget committed:
clang format

tewinget committed:
correctly handle/decrypt path control responses

tewinget committed:
actually handle session init messages...

tewinget committed:
actually respond to session init...

tewinget committed:
log message fixes

tewinget committed:
actually send session_init

turns out if you queue the session_init packet until the session
is initialized, neither chicken nor egg can happen...

tewinget committed:
missing returns on failure case

tewinget committed:
fix session tag bt decode on session init

tewinget committed:
assign aligned hopid on outbound session path creation

tewinget committed:
re-implement session path switch and publish cc

also remove dead code from the previous versions of those

tewinget committed:
check hopid makes at least some sense on path switch

tewinget committed:
session close reimplemented

tewinget committed:
replace span::first (possibly buggy on apple?)

tewinget committed:
span::size is seemingly just broken on apple?

tewinget committed:
Fix rare off-by-one in inbound path build logic

There was an incredibly rare case where SessionEndpoint::update_paths
could get past the "we need more paths" check because of a timing issue
(not passing "now" to num_paths()), and then the code would index an
array at -1, which would corrupt the stack in a way that led to some
crashing.  This is fixed by passing "now" correctly to that call and
correcting a conditional later which was simply wrong/useless anyway.

In addition, the other reason this corruption was possible is that the
code was trying to not build a path in the first "time bucket" (e.g. 0-5
min expiration), but this was implemented in a way which has a few
undesirable side-effects.  Instead, we now *do* build a path in that
"bucket" if it doesn't have one and should.

tewinget committed:
default embedded port to 0

we may want to consider this for all clients, but definitely for embedded

jagerman commented on llarp/crypto/crypto.hpp:
```diff
@@ -20,6 +22,19 @@ namespace llarp::crypto
+
+    /// encrypts a buffer in-place, putting a MAC in the final bytes
+    void xchacha20_poly1305_encrypt(std::span<std::byte> buf, const SharedSecret& secret, const SymmNonce& nonce);
+    void xchacha20_poly1305_encrypt(std::vector<std::byte>& buf, const SharedSecret& secret, const SymmNonce& nonce);
```
This second overload is unnecessary: a `std::vector<std::byte>` is already implicitly convertible to a mutable span.

jagerman commented on llarp/crypto/crypto.hpp:
```diff
@@ -20,6 +22,19 @@ namespace llarp::crypto
 
+    static constexpr size_t MAC_SIZE = crypto_aead_xchacha20poly1305_ietf_ABYTES;
+
+    /// encrypts a buffer in-place, putting a MAC in the final bytes
```
```suggestion
    /// encrypts a buffer in-place, putting a MAC in the final bytes (which must be allocated
    /// in the span but are not part of the data that is actually encrypted).
```

tewinget committed:
remove redundant overload, fixup comment, move constant

tewinget committed:
bump libquic

tewinget committed:
libquic stream changes, tcp tunnel still broken so ignore for now

tewinget committed:
fix off-by-one introduced by fixing off-by-one

The previous off-by-one fix moved the off-by-one a bit later, causing attempts
to build paths which expire in the past.
